### PR TITLE
added tbib.org scraper

### PR DESCRIPTION
**Pull request by LainLayer**
It's not like i want to make a pull request or anything, b-baka!

### CODE REVIEW

File: scrappers/tbib/lib/Tbib.rb [hunk 1]
```diff
@@ -0,0 +1,76 @@
+class Tbib
+	include RethinkDB::Shortcuts
+
+	def initialize()
+		@conn = r.connect(:host => $config.server_ip, :port => $config.server_port)
+		$log.log 'connected to database!'
+		@conn.use($config.database)
+		@mutex = Mutex.new
```
Comment (vaartis) [inline scrappers/tbib/lib/Tbib.rb]: This should probably be a semaphore (look into the `concurrent-ruby` gem) with a configureable limit (something like `pages-thread-limit` in your config.toml's `[other]`)

File: scrappers/tbib/lib/Tbib.rb [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+		if !r.table_list.run(@conn).include?($config.table)
+			r.table_create($config.table).run(@conn)
```
Comment (vaartis) [inline scrappers/tbib/lib/Tbib.rb]: The table should use the image `id` as primary key, you can read how to do that [here](https://www.rethinkdb.com/api/ruby/table_create/)

File: scrappers/tbib/lib/Tbib.rb [hunk 3]
```diff
@@ -0,0 +1,76 @@
+class Tbib
+	include RethinkDB::Shortcuts
+
+	def initialize()
+		@conn = r.connect(:host => $config.server_ip, :port => $config.server_port)
+		$log.log 'connected to database!'
+		@conn.use($config.database)
```
Comment (vaartis) [inline scrappers/tbib/lib/Tbib.rb]: You should check if the database exists and create it if not, there's a chance it doesn't exist just like the table

File: scrappers/tbib/lib/Tbib.rb [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+			exit 0
+		end
+		images = []
+		procs = []
+		(0..(c/100 + (c%100 == 0 ? 0 : 1))).to_a.each do |pid|
+			procs << proc {
+				Thread.new do
+					url = "https://tbib.org/index.php?page=dapi&s=post&q=index&tags=#{$config.tags.join('+')}&json=1&pid=#{pid}"
+					j = JSON.parse(open(url).read)
+					@mutex.synchronize do
```
Comment (vaartis) [inline scrappers/tbib/lib/Tbib.rb]: Should probably move that up a little, so it'd also handle the request pausing (and also change that to a semaphore, as stated previously)

File: scrappers/tbib/scraper.rb [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+
+OptionParser.new do |ops|
+	ops.banner = 'Usage: ruby scraper.rb [options]'
+
+	ops.on('-h', '--help', 'Prints this text.') do
+		puts ops
+		exit 0
+	end
+
+	ops.on('-d', '--default-config', 'Makes the scraper use the default configuration options.') do
```
Comment (vaartis) [inline scrappers/tbib/scraper.rb]: This should generate and dump the default `config.toml` to the stdout, not make the scrapper use the default config

**Review by vaartis — CHANGES_REQUESTED** [review]
- [ ] The `config.toml` file should not exist just like that (default config needs to be printedt with the `--default-config`)
- [ ] Scrapper should create an entry in the `scrapper_info` table with the information about it (refer to the specification and the [safeboory scrapper](https://github.com/vaartis/imageboard-indexer/blob/ae7a5426a25aaa9f48a3c69e82068a421ed4cd40/scrappers/safebooru-py/scrapper.py#L212)
- [ ] There should be a blacklisted-tags options as an opposite of tags
- [ ] There should be a general.start option that specifies how should the frontend run the scrapper (something like `ruby __file__`), it should be written to the `scrapper_info` table (as the specification says).
- [ ] The scrapper lacks image downloading and things that are connected to it (like the `images.save-path` option, specifying where to save the images).
- [ ] Would be nice to have some comments and documentation
- [ ] Maybe change `gems.txt` to a `Gemfile`

File: scrappers/tbib/lib/Tbib.rb (hunk 1 above)
Comment (LainLayer) [reply scrappers/tbib/lib/Tbib.rb]: semaphore not needed. threads are limited here:
```
while !procs.empty?
  $log.progress "Downloading images...", 1.0 - procs.length.to_f/ol
  b = procs.shift(8)
  b.each_with_index { |d, i| b[i] = d.call }
  b.each { |d| d.join }
  sleep 1
end
```

it takes 8 threads, starts them, and waits for them to finish, then starts the next 8.

Comment (vaartis) [reply scrappers/tbib/lib/Tbib.rb]: You should probably put a comment there explaining this behaviour

File: scrappers/tbib/lib/Tbib.rb (hunk 2 above)
Comment (LainLayer) [reply scrappers/tbib/lib/Tbib.rb]: done ;)
`r.table_create($config.table, :primary_key => 'id').run(@conn)`

File: scrappers/tbib/lib/Tbib.rb (hunk 3 above)
Comment (LainLayer) [reply scrappers/tbib/lib/Tbib.rb]:
```
if !r.db_list().run(@conn).include?($config.database)
      r.db_create($config.database)
end
```
:D

Comment (LainLayer) [reply scrappers/tbib/lib/Tbib.rb]: (and with a `.run(@conn)` )

**Comment by vaartis** [comment]
:monocle_face:

**Pull request merged**